### PR TITLE
refactor prepare_mask_and_masked_image with VaeImageProcessor

### PR DESCRIPTION
**Pull request by yiyixuxu**
first attempt to refactor inpainting pipelines with VaeImageProcessor
lots of tests need to be added

This example work as expected after refactoring 

```python
import PIL
import requests
import torch
from io import BytesIO

from diffusers import StableDiffusionInpaintPipeline


def download_image(url):
    response = requests.get(url)
    return PIL.Image.open(BytesIO(response.content)).convert("RGB")


img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = download_image(img_url).resize((512, 512))
mask_image = download_image(mask_url).resize((512, 512))

generator = torch.Generator(device="cuda").manual_seed(0)

pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16,
)
pipe = pipe.to("cuda")

prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
image = pipe(prompt=prompt, image=init_image, mask_image=mask_image, generator=generator).images[0]
image.save("yellow_cat.png")
```

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 1]
```diff
@@ -858,9 +866,17 @@ def __call__(
...
-            image, mask_image, height, width, return_image=True
-        )
+
+        init_image = self.image_processor.preprocess(image, height=height, width=width)
+        init_image = init_image.to(dtype=torch.float32)
+
+        mask = self.mask_processor.preprocess(mask_image, height=height, width=width)
+        # binarize mask
+        mask[mask < 0.5] = 0
+        mask[mask > 0.5] = 1
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:876]:
```suggestion
        mask[mask < 0.5] = 0
        mask[mask > 0.5] = 1
```

shouldn't we do this in the preprocess function?

File: src/diffusers/image_processor.py [hunk 2]
```diff
@@ -52,8 +52,16 @@ def __init__(
         resample: str = "lanczos",
         do_normalize: bool = True,
         do_convert_rgb: bool = False,
+        do_convert_grayscale: bool = False,
```
Comment (patrickvonplaten) [inline src/diffusers/image_processor.py:55]: nice idea!

File: src/diffusers/image_processor.py [hunk 3]
```diff
@@ -52,8 +52,16 @@ def __init__(
...
     ):
         super().__init__()
+        if do_convert_rgb and do_convert_grayscale:
+            warnings.warn(
```
Comment (patrickvonplaten) [inline src/diffusers/image_processor.py:59]: maybe better to throw an error here actually

**Comment by patrickvonplaten** [comment]
Very nice first draft! cc @sayakpaul @pcuenca @williamberman here for a review as well

File: src/diffusers/image_processor.py [hunk 4]
```diff
@@ -51,9 +51,18 @@ def __init__(
...
         do_convert_rgb: bool = False,
+        do_convert_grayscale: bool = False,
     ):
         super().__init__()
+        if do_convert_rgb and do_convert_grayscale:
+            raise ValueError(
+                "`do_convert_rgb` and `do_convert_grayscale` can not both be set to `True`,"
+                " if you intended to convert the image into RGB format, please set `do_convert_grayscale = False`.",
+                " if you intended to convert the image into grayscale format, please set `do_convert_rgb = False`",
+            )
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:64]: Very descriptive! Looks good.

File: src/diffusers/image_processor.py [hunk 5]
```diff
@@ -119,31 +128,84 @@ def denormalize(images):
...
+
+        Args:
+            image(`PIL.Image.Image`, `np.ndarray` or `torch.Tensor`):
+                the image input, can be a PIL image, numpy array or pytorch tensor. if it is a numpy array, should have
+                shape [batch, height, width] or [batch, height, width, channel] if it is a pytorch tensor, should have
+                shape [batch, channel, height, width]
+            height (`int`, *optional*, defaults to `None`):
+                The height in preprocessed image. If `None`, will use the height of `image` input
+            width (`int`, *optional*`, defaults to `None`):
+                The width in preprocessed. If `None`, will use the width of the `image` input
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:164]: Let's maintain casing:

* "the" -> "The" when it's being placed at the beginning. 
* Fullstops to end the sentences.

File: src/diffusers/image_processor.py [hunk 6]
```diff
@@ -119,31 +128,84 @@ def denormalize(images):
...
         image = image.resize((width, height), resample=PIL_INTERPOLATION[self.config.resample])
         return image
 
+    def binarize(self, image: PIL.Image.Image) -> PIL.Image.Image:
+        """
+        create a mask
+        """
+        image[image < 0.5] = 0
+        image[image >= 0.5] = 1
+        return image
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:207]: Should 0 and 1 be registered into the config vars?

File: src/diffusers/image_processor.py [hunk 7]
```diff
@@ -154,6 +216,25 @@ def preprocess(
...
+                image = image.unsqueeze(1)
+            else:
+                # if it is a numpy array, it could have 2 possible shapes:
+                #   1. batch x height x width: insert channel dimension on last position
+                #   2. height x width x channel: insert batch dimension on first position
+                if image.shape[-1] == 1:
+                    image = np.expand_dims(image, axis=0)
+                else:
+                    image = np.expand_dims(image, axis=-1)
+
```
Comment (sayakpaul) [inline src/diffusers/image_processor.py:237]: For easier operability, does it make sense to first convert the input tensor to a NumPy array and then operate from there?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 8]
```diff
@@ -858,9 +866,14 @@ def __call__(
...
-        mask, masked_image, init_image = prepare_mask_and_masked_image(
-            image, mask_image, height, width, return_image=True
-        )
+
+        init_image = self.image_processor.preprocess(image, height=height, width=width)
+        init_image = init_image.to(dtype=torch.float32)
+
+        mask = self.mask_processor.preprocess(mask_image, height=height, width=width)
+
+        masked_image = init_image * (mask < 0.5)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:875]: Register 0.5 as a config var?

**Review by sayakpaul — COMMENTED** [review]
Looking quite good. 

Let's try to think of a non-exhaustive list of the scenarios we need to think of to ensure this refactor is robust and write test cases for each of them.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:875]: Do you mean of the init of the VAEProcessor or the vae model config ? IMO it's general enough to not have to be registered in a config

File: src/diffusers/image_processor.py (hunk 7 above)
Comment (yiyixuxu) [reply src/diffusers/image_processor.py:237]: @sayakpaul 

Our logic to convert the PyTorch tensor into a NumPy array assumes a 3D tensor has shape `channel x height x width`, which doesn't apply to grayscale images

for example if we have a tensor with shape `(5, 256,256)`, if we try to convert it into numpy with current logic directly, we will do this:
 `(5, 256,256)` -> put it into a list :`[ (5,256,256)]` -> torch.stack the list when not 4d: `(1, 5, 256, 256)`

Comment (sayakpaul) [reply src/diffusers/image_processor.py:237]: Why cannot we add a dummy channel to represent the grayscale images then? Just trying to understand it better.

Comment (yiyixuxu) [reply src/diffusers/image_processor.py:237]: @sayakpaul 
yeah and that's what this section of code is doing: 
Adding a dummy channel to represent either a missing channel or batch dimension to remove the ambiguity

**Comment by yiyixuxu** [comment]
@patrickvonplaten 
should we unify the default behavior for when the user does not pass the `height` and `width` arguments?

in SD inpaint pipeline, when `height` and `width` is `None`, it will default to `sample_size * vae scale factor`;
in controlnet inpaint, it will resize to the closest multiple of `8`, which I think makes much more sense

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py#L866

File: src/diffusers/image_processor.py [hunk 9]
```diff
@@ -164,42 +249,47 @@ def preprocess(
...
-                    f"Currently we only support resizing for PIL image - please resize your pytorch tensor to be divisible by {self.config.vae_scale_factor}"
-                    f"currently the sizes are {height} and {width}. You can also pass a PIL image instead to use resize option in VAEImageProcessor"
+                    f"Currently we only support resizing for PIL image - please resize your torch tensor to be {height} and {width}"
+                    f"currently the sizes are {image.shape[2]} and {image.shape[3]}. You can also pass a PIL image instead to use resize option in VAEImageProcessor"
                 )
 
         # expected range [0,1], normalize to [-1,1]
         do_normalize = self.config.do_normalize
-        if image.min() < 0:
+        if image.min() < 0 and do_normalize:
```
Comment (yiyixuxu) [inline src/diffusers/image_processor.py:292]: if user configured the image_processor to have `do_normalize=False`, the expected range should be `[-1,1]` and we shouldn't send a warning (this is the case for controlnet image)

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py [hunk 10]
```diff
@@ -943,8 +925,22 @@ def _encode_vae_image(self, image: torch.Tensor, generator: torch.Generator):
...
-        image: Union[torch.Tensor, PIL.Image.Image] = None,
-        mask_image: Union[torch.Tensor, PIL.Image.Image] = None,
+        image: Union[
+            torch.FloatTensor,
+            PIL.Image.Image,
+            np.ndarray,
+            List[torch.FloatTensor],
+            List[PIL.Image.Image],
+            List[np.ndarray],
+        ] = None,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py:935]:
```suggestion
        image: Union[
            torch.FloatTensor,
            PIL.Image.Image,
            np.ndarray,
            List[torch.FloatTensor],
            List[PIL.Image.Image],
            List[np.ndarray],
        ] = None,
```
might be worth to define a new type here actually

**Review by patrickvonplaten — COMMENTED** [review]
Looks cool! Can we apply the changes also right away to SDXL and SDXL controlnet?

**Comment by yiyixuxu** [comment]
>  Can we apply the changes also right away to SDXL and SDXL controlnet?

updated SDXL-inpainting

SDXL controlnet does not have masks input, and we already used a image_processor to process the control condition:) we've added Image_processor to all the other SDXL pipelines wherever there is image input

**Comment by yiyixuxu** [comment]
don't now why this test fail - it passes on my machine and this PR did not touch the paint_by_example pipeline at all https://github.com/huggingface/diffusers/actions/runs/5959195493/job/16165075058?pr=4444#step:6:13213

**Review by patrickvonplaten — APPROVED** [review]
Very nice! Let's merge it before we run into ugly merge conflicts :-)
